### PR TITLE
Env Awareness & Better Retries in Staging

### PR DESCRIPTION
**Pull request by dccourtney**
risk: low-medium

**Deploy Master**: please update [aquaMantis](https://github.com/iGoDigital-LLC/aquaMantis/blob/master/Gemfile) to the release tag after this is deployed to other apps

Resulted out of testing PI_SESSION ticket - [W-3392494](https://gus.my.salesforce.com/a07B0000002RgsFIAS)

- Add config that is environment aware
- In staging - try qa3s1 auth url then fallback to qa1s1 
- All other envs use normal auth url

@iGoDigital-LLC/alot

### CODE REVIEW

File: lib/fuelsdk/client.rb [hunk 1]
```diff
@@ -108,32 +109,52 @@ def initialize(params={}, debug=false)
...
-			#If we don't already have a token or the token expires within 5 min(300 seconds)
-			if (self.access_token.nil? || Time.new + 300 > self.auth_token_expiration || force) then
-			payload = Hash.new.tap do |h|
-				h['clientId']= id
-				h['clientSecret'] = secret
-				h['refreshToken'] = refresh_token if refresh_token
-				h['accessType'] = 'offline'
-			end
+		def get_token
+			retry_call = FuelSDK::Config.staging?
```
Comment (mjrust) [inline lib/fuelsdk/client.rb]: Why did they originally use 8 spaces? Should you fix it or let it be?

Comment (jkieffaber) [inline lib/fuelsdk/client.rb]: I would leave it as it.  The less you can touch this gem the better.

**Comment by jonathan** [comment]
👍

**Comment by jkieffaber** [comment]
@dccourtney is this still on hold?

**Comment by dccourtney** [comment]
yes, still on hold. ill remove the deploy label for now. there seems to be a dependency issue with rack that needs to be resolved. doesn't seem to be related to these changes.

**Comment by jkieffaber** [comment]
Looks good to me.  @adamjgray Can you take a look at this?

File: lib/new.rb [hunk 2]
```diff
@@ -223,47 +225,64 @@ def debug=(value)
...
-          if @endpoint.nil? then
+          if @endpoint.nil?
             self.determineStack
           end
 
           @authObj = {'oAuth' => {'oAuthToken' => @internalAuthToken}}
           @authObj[:attributes!] = { 'oAuth' => { 'xmlns' => 'http://exacttarget.com' }}
 
-          myWSDL = File.read(@path + '/ExactTargetWSDL.xml')
+          wsdl = File.read(@path + '/ExactTargetWSDL.xml')
```
Comment (mediocretes) [inline lib/new.rb]: Consider string interpolation, if you're classy, or File.join, if you want to be safe.

File: lib/fuelsdk/client.rb (hunk 1 above)
Comment (adamjgray) [inline lib/fuelsdk/client.rb]: If the tabs hurt your eyes: https://github.com/iGoDigital-LLC/FuelSDK-Ruby/pull/4/files?ts=2

File: lib/new.rb [hunk 3]
```diff
@@ -223,47 +225,64 @@ def debug=(value)
...
-          @authTokenExpiration = Time.new + tokenResponse['expiresIn']
-          @internalAuthToken = tokenResponse['legacyToken']
-          if tokenResponse.has_key?("refreshToken") then
-            @refreshKey = tokenResponse['refreshToken']
+      token_response
+    end
+
+    def refreshToken(force = nil)
+      #If we don't already have a token or the token expires within 5 min(300 seconds), get one
+      if (@authToken.nil? || Time.new + 300 > @authTokenExpiration) || force
```
Comment (mediocretes) [inline lib/new.rb]: What about

```
if force || @auth_token.nil? || auth_token_expired?
```

And then write auth_token_expired? for clarity?

**Pull request closed without merging**